src/charm.py

```diff
@@ -26,20 +28,89 @@ def __init__(self, *args: typing.Any) -> None:
         super().__init__(*args)
         self.framework.observe(self.on.config_changed, self.config_service)
 
+    def flask_container(self, require_connected: bool = True) -> Container:
+        """Get the flask application workload container controller.
+
+        Args:
+            require_connected: if set to ``True``, a runtime exception will be raised if the
+                pebble inside the container is not ready.
+
+        Return:
+            The controller of the flask application workload container.
+
+        Raises:
+            RuntimeError: if the pebble inside  the container is not ready while the
+                ``require_connected`` is set to True.
+        """
+        container = self.unit.get_container("flask-app")
+        if require_connected and not container.can_connect():
+            raise RuntimeError("pebble inside flask-app container is not ready")
+        return container
+
     def config_service(self, event: ConfigChangedEvent) -> None:
         """Configure the flask pebble service layer.
 
         Args:
             event: the config-changed event that trigger this callback function.
         """
-        container = self.unit.get_container("flask-app")
+        container = self.flask_container(require_connected=False)
         if not container.can_connect():
             event.defer()
             return
+        service_name = "flask-app"
+        gunicorn_config_path = "/srv/flask/gunicorn.conf.py"
+        current_gunicorn_config = self.pull_file(gunicorn_config_path)
```

Comment:
I would no recommend pulling the file, the event should indicate if the change requires a reload
For example any config change that modify a property starting with "webserver" will require a reload, this will be known in the config_changed event
Then the config_changed event will call the service to generate the file passing the state to it. I think this service could be a "gunicorn_service" that will know where to put the config file, how to generate it, and how to reload the service, and how to know if the reloading was successful

Reply:
I have refactored the charm to conform with the new standard, thanks!

Reply:
Ok let me review the current version